### PR TITLE
Enable in oss

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124031
* #122825
* #123229
* #122642
* #124030

Biggest movement is 4% HF inference, 9% TIMM inference. Note, this is max-autotune mode so we are more tolerant of compilation increases. We could improve compilation time by limiting:
```
# Take how many of the top triton kernels to benchmark epilogue
max_epilogue_benchmarked_choices = 3
```

There is a hf_Whisper failure which you can repro on main without this stack with `TORCHINDUCTOR_MAX_AUTOTUNE_GEMM_BACKENDS=TRITON TORCHINDUCTOR_MAX_AUTOTUNE=1 python benchmarks/dynamo/torchbench.py --backend inductor --amp --accuracy --training --only hf_Whisper`. When you turn off epilogue fusion, it fixes the accuracy. I bisected the failure to an epilogue, however when you compare the results of that epilogue with the corresponding separate kernels the results of the output are equivalent.

Inference:

image

Training:

Screenshot 2024-04-16 at 6 16 30 PM

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by shunting314** [comment]
Is there a perf run link?

I'm also curious about other metrics besides speedup.

**Comment by eellison** [comment]
@shunting314 added

**Comment by shunting314** [comment]
Besides hf_Whisper, I saw that hf_Reformer and phlippe_resnet also start failing. If they are all accuracy failures and we want to raise tolerance, should we fix them now. Debugging/fixing them right now or cutting an issue to track should make it easier since we can trace back to the cause.

Besides that, nice perf wins! I'll defer to other reviewers if this is a good trade off between compilation time (in max-autotune mode) and perf wins.

**Comment by eellison** [comment]
Filed issue here for the resnet thing: https://github.com/openai/triton/issues/3684.

Will add a try catch in epilogue (similar to what exists [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codecache.py#L318-L321) for base template).

**Comment by eellison** [comment]
Investigated hf_Reformer. I was able to repro the accuracy failure through a single epilogue. Adding a 
`to(tl.float16).to(tl.float32)` fixed the accuracy failure. This is more faithful to eager numerics even though it is worse precision. 

This is an example of where adding a "faithful to eager" (with additional downcasts) mode might have been helpful.

**Review by Chillee — APPROVED** [review]

**Review by shunting314 — APPROVED** [review]
